### PR TITLE
WIP: neural network demo

### PR DESCRIPTION
**Pull request by oxinabox**
A very simple neural network example.

When I am done it will  probably either be online, or full batch, vanilla gradient desent.
For the XOR problem, and for FashionMNIST

### CODE REVIEW

File: examples/NeuralNetwork/create_data.jl [hunk 1]
```diff
@@ -0,0 +1,41 @@
+using Pkg: Pkg
```
Comment (oxinabox) [inline examples/NeuralNetwork/create_data.jl]: This file is for generating the FashionMNIST data in a form that can be copy-pasted or linked in to a Dex file

Comment (oxinabox) [reply examples/NeuralNetwork/create_data.jl]: see #40

Comment (oxinabox) [reply examples/NeuralNetwork/create_data.jl]: see #40 
Previously i had the genrated code for the constants pasted in to the main file
but it made refreshing the web view take like multiple seconds while it reparsed the file
and github refused to render it

File: examples/NeuralNetwork/neural_network.dx [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+xorTrainTargets
+
+(filter trueXorClass xorTrainInputs).(asidx 1)
+
+' ## Xor Net
+This is a two layer net.
+AFAICT no data structure in Dex right now can hold a arbitrary collection
+of neural network layers/parameters, as it would be a heterogeneous vector, or a tuple of non-fixed length
+As such need different functions for each depth of network
```
Comment (oxinabox) [inline examples/NeuralNetwork/neural_network.dx]: This is a problem.
Maybe there is a way to do Hetrogenious Vectors?

File: examples/NeuralNetwork/neural_network.dx [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+apply2Net:: ((hz=>(iz=>Real), hz=>Real), (oz=>(hz=>Real), oz=>Real)) -> iz=>Real -> oz=>Real
+apply2Net (layer1, layer2) input = 
+    sigmoid $ applyAffine layer2 $ sigmoid $ applyAffine layer1 input
+
+apply2Net xorNetParams [0.9, 0.1]
+
+(binaryCrossEntropy [1.0]) $ apply2Net xorNetParams [0.9, 0.1]
+
+eval2NetLoss:: (oz=>Real->oz=>Real->Real) -> oz=>Real -> iz=>Real -> ((hz=>(iz=>Real), hz=>Real), (oz=>(hz=>Real), oz=>Real)) -> Real
```
Comment (oxinabox) [inline examples/NeuralNetwork/neural_network.dx]: This is really complex
See #48

File: examples/NeuralNetwork/neural_network.dx [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+eval2NetLoss loss target input netParams = ..
+    loss target $ apply2Net netParams input
+
+eval2NetLoss binaryCrossEntropy [1.0] [0.9, 0.1] xorNetParams 
+
+xorLoss:: ((hz=>(2=>Real), hz=>Real), (1=>(hz=>Real), 1=>Real)) -> Real
+xorLoss = eval2NetLoss binaryCrossEntropy [1.0] [0.9, 0.1]
+
+
+grad xorLoss xorNetParams
```
Comment (oxinabox) [inline examples/NeuralNetwork/neural_network.dx]: this currently errors:


```
Zeros not implemented for Int
CallStack (from HasCallStack):
  error, called at src/lib/Embed.hs:189:13 in dex-0.1.0.0-9pQ6aTLha3fLdCjkbXPWRy:Embed
```

Comment (oxinabox) [reply examples/NeuralNetwork/neural_network.dx]: I am hoping `grad` on a **Tuple of Tuples of Arrays**, works and returns  a **Tuple of Tuples of Arrays** of derivatives.
Or equivelent.

I think its important to support differentiation on composite structures.
This is the thing ChainRules.jl calls a `Composite` differential,
and Swift calls a Tangent Associated Type.

Comment (oxinabox) [reply examples/NeuralNetwork/neural_network.dx]: Seems like this kinda thing works:
```
>=> foo1:: (2=>Real, 2=>Real)->Real
... foo1 (xs,ys) = sum for i. xs.i*ys.i

>=> foo ([1.0, 2.0], [3.0, 4.0])
Error: variable not in scope: foo
>=> foo1 ([1.0, 2.0], [3.0, 4.0])
11.0
>=> grad foo1
Type error: Can't print values of type: (((2=>Real), (2=>Real)) -> ((2=>Real), (2=>Real)))
>=> grad foo1 ([1.0, 2.0], [3.0, 4.0])
([3.0, 4.0], [1.0, 2.0])
```

**Comment by oxinabox** [comment]
Checkout that last commit.
So much more readale with parametric type aliases

**Comment by oxinabox** [comment]
I was intending to come back to this at some point,
just been really busy

**Comment by dougalm** [comment]
Sorry, I didn't close this intentionally! I think it was just a side effect of renaming the default branch. Feel free to re-open any time.

**Comment by oxinabox** [comment]
I think I will start a new NN demo from scratch

**Pull request closed without merging**